lang/java/reef-runtime-multi/src/main/java/org/apache/reef/runtime/multi/client/AbstractDriverConfigurationProvider.java

```diff
@@ -0,0 +1,119 @@
+import org.apache.reef.tang.Configuration;
+import org.apache.reef.tang.Configurations;
+import org.apache.reef.tang.formats.AvroConfigurationSerializer;
+import org.apache.reef.tang.formats.ConfigurationModule;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.net.URI;
+import java.util.ArrayList;
+
+/**
+ * Provides base class for driver configuration providers for multi runtimes.
+ */
+abstract class AbstractDriverConfigurationProvider implements DriverConfigurationProvider {
+
+  private static final String CHARSET_NAME = "ISO-8859-1";
+
+  protected static String serializeConfiguration(final ConfigurationModule configModule) {
+    final Configuration localDriverConfiguration = configModule.build();
+    final AvroConfigurationSerializer serializer = new AvroConfigurationSerializer();
+    return serializer.toString(localDriverConfiguration);
+  }
+
+  protected static String serializeRuntimeDefinition(final String serializedConfig,
+                                                     final boolean defaultRuntime,
+                                                     final String runtimeName) {
+    RuntimeDefinition rd =
+            new RuntimeDefinition(runtimeName, serializedConfig, defaultRuntime);
+    final DatumWriter<RuntimeDefinition> configurationWriter =
+            new SpecificDatumWriter<>(RuntimeDefinition.class);
+    final String serializedConfiguration;
+    try (final ByteArrayOutputStream out = new ByteArrayOutputStream()) {
+      final JsonEncoder encoder = EncoderFactory.get().jsonEncoder(RuntimeDefinition.SCHEMA$, out);
+      configurationWriter.write(rd, encoder);
+      encoder.flush();
+      out.flush();
+      serializedConfiguration = out.toString(CHARSET_NAME);
+    } catch (final IOException e) {
+      throw new RuntimeException(e);
+    }
+
+    return serializedConfiguration;
+  }
+
+  private static Configuration generateConfigurationModule(final ArrayList<String> serializedConfigurations,
+                                                           final String jobId,
+                                                           final String clientRemoteId) {
+    ConfigurationModule module = MultiRuntimeDriverConfiguration.CONF
+            .set(MultiRuntimeDriverConfiguration.JOB_IDENTIFIER, jobId)
+            .set(MultiRuntimeDriverConfiguration.CLIENT_REMOTE_IDENTIFIER, clientRemoteId);
+    for (String config : serializedConfigurations) {
+      module = module.set(MultiRuntimeDriverConfiguration.SERIALIZED_RUNTIME_DEFINITION, config);
+    }
+
+    return module.build();
+  }
+
+  /**
+   * Generates needed driver configuration modules.
+   *
+   * @param jobFolder      the job folder
+   * @param clientRemoteId the client remote id
+   * @param jobId          the job id
+   * @return array of configuration modules for runtimes, when the first element contains the default runtime
+   */
+  protected abstract String[] getDriverConfiguration(final URI jobFolder,
```

Comment:
Why `String[]` as the return value and not `Collection<Configuration>`? Or `Iterable<Configuration>`? The serialization into strings seems like a implementation detail which currently leaks.